File: checks/python/src/section508/checks/decorative_image_silent.py
from section508.dom import SPACE, attr, has_attr, select


def check(root):
    if not root:
        return []

    images = select(root, "img")
    if not images:
        return ["This section has no decorative image."]

    for image in images:
        if not has_attr(image, "alt"):
            return ["This decorative image is missing the alt attribute."]
        alt = attr(image, "alt")
        if alt != "":
            shown = SPACE.sub(" ", alt or "").strip() or "text"
            return [f"“{shown}” will be announced. Decorative images need empty alt."]
    return []

File: checks/ruby/lib/section508/checks/decorative_image_silent.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Ornament is silent. Missing alt is not empty alt.
    class DecorativeImageSilent
      def self.call(root)
        return [] unless root
        return [] unless root.respond_to?(:css)

        images = Dom.elements(root, "img")
        return ["This section has no decorative image."] if images.empty?

        images.each do |image|
          return ["This decorative image is missing the alt attribute."] unless Dom.has_attr?(image, "alt")

          alt = image["alt"]
          next if alt == ""

          shown = alt.to_s.gsub(/\s+/, " ").strip
          shown = "text" if shown.empty?
          return ["“#{shown}” will be announced. Decorative images need empty alt."]
        end
        []
      end
    end
  end
end
